File: frontend/src/services/dateUtils.js
/**
 * dateUtils.js - Utility functions for handling dates consistently throughout the application
 * Handles timezone-aware conversions and formatting to prevent scheduling issues
 */

/**
 * Formats a date to ISO date string (YYYY-MM-DD) without time component
 * @param {Date} date - The Date object to format
 * @returns {string} Formatted date string in YYYY-MM-DD format
 */
export const formatDateToISODate = (date) => {
    if (!(date instanceof Date)) {
      throw new Error('Invalid date object provided to formatDateToISODate');
    }
    
    const year = date.getFullYear();
    const month = (date.getMonth() + 1).toString().padStart(2, '0');
    const day = date.getDate().toString().padStart(2, '0');
    
    return `${year}-${month}-${day}`;
  };
  
  /**
   * Formats a date to a full ISO datetime string but preserves local time
   * by excluding timezone offset (YYYY-MM-DDTHH:MM:SS)
   * @param {Date} date - The Date object to format
   * @returns {string} Formatted datetime string in YYYY-MM-DDTHH:MM:SS format
   */
  export const formatToLocalISOString = (date) => {
    if (!(date instanceof Date)) {
      throw new Error('Invalid date object provided to formatToLocalISOString');
    }
    
    const year = date.getFullYear();
    const month = (date.getMonth() + 1).toString().padStart(2, '0');
    const day = date.getDate().toString().padStart(2, '0');
    const hours = date.getHours().toString().padStart(2, '0');
    const minutes = date.getMinutes().toString().padStart(2, '0');
    const seconds = date.getSeconds().toString().padStart(2, '0');
    
    // Format as local ISO string without timezone to preserve local time
    return `${year}-${month}-${day}T${hours}:${minutes}:${seconds}`;
  };
  
  /**
   * Parses a date string to a Date object, handling various formats consistently
   * @param {string} dateStr - The date string to parse
   * @returns {Date|null} Parsed Date object or null if invalid
   */
  export const parseDate = (dateStr) => {
    if (!dateStr) return null;
    
    try {
      // Handle ISO format with timezone (convert to local time correctly)
      if (typeof dateStr === 'string' && dateStr.includes('Z')) {
        return new Date(dateStr);
      }
      
      // Handle ISO format without timezone (interpret as local time)
      if (typeof dateStr === 'string' && dateStr.includes('T')) {
        const [datePart, timePart] = dateStr.split('T');
        const [year, month, day] = datePart.split('-').map(Number);
        const [hours, minutes, seconds] = timePart.split(':').map(Number);
        
        // Construct date using local timezone
        return new Date(year, month - 1, day, hours, minutes, seconds || 0);
      }
      
      // Handle YYYY-MM-DD format (set to noon local time to avoid timezone issues)
      if (typeof dateStr === 'string' && dateStr.match(/^\d{4}-\d{2}-\d{2}$/)) {
        const [year, month, day] = dateStr.split('-').map(Number);
        return new Date(year, month - 1, day, 12, 0, 0);
      }
      
      // Default parsing for other formats
      const parsedDate = new Date(dateStr);
      return isNaN(parsedDate.getTime()) ? null : parsedDate;
    } catch (error) {
      console.error('Error parsing date:', error);
      return null;
    }
  };
  
  
  /**
   * Safely handles date inputs for API requests
   * @param {Object} data - Object containing date fields
   * @param {string} fieldName - Name of the date field to process
   * @returns {Object} Modified data object with properly formatted date
   */
  export const prepareScheduledDateForApi = (data, fieldName = 'scheduled_date') => {
    if (!data || !data[fieldName]) return data;
    
    const result = { ...data };
    
    // Handle different input types
    if (result[fieldName] instanceof Date) {
      // For estimate scheduling with time component
      if (data.hour !== undefined || data.minute !== undefined) {
        // Use custom date/time from inputs
        const hours = data.hour !== undefined ? data.hour : result[fieldName].getHours();
        const minutes = data.minute !== undefined ? data.minute : result[fieldName].getMinutes();
        
        // Create a new date with the specified time
        const dateWithTime = new Date(result[fieldName]);
        dateWithTime.setHours(hours, minutes, 0, 0);
        
        // Format as local ISO string
        result[fieldName] = formatToLocalISOString(dateWithTime);
      } else if (data.includeTime === true) {
        // Include full time information
        result[fieldName] = formatToLocalISOString(result[fieldName]);
      } else {
        // For job scheduling where we only need the date
        result[fieldName] = formatDateToISODate(result[fieldName]);
      }
    } else if (typeof result[fieldName] === 'string' && result[fieldName].match(/^\d{4}-\d{2}-\d{2}$/)) {
      // It's already in YYYY-MM-DD format
      // Add time component if needed
      if (data.hour !== undefined && data.minute !== undefined) {
        const hour = data.hour.toString().padStart(2, '0');
        const minute = data.minute.toString().padStart(2, '0');
        result[fieldName] = `${result[fieldName]}T${hour}:${minute}:00`;
      }
    }
    
    return result;
  };


  // dateUtils.js - Create this file for consistent date handling
import { parse, format, parseISO } from 'date-fns';

/**
 * Converts a date to an ISO string without timezone information
 * to preserve the exact local time when sending to the server
 * @param {Date} date - The date to convert
 * @returns {string} ISO date string without timezone info
 */
export const toLocalISOString = (date) => {
  if (!(date instanceof Date)) return null;
  
  // Extract local date components
  const year = date.getFullYear();
  const month = String(date.getMonth() + 1).padStart(2, '0');
  const day = String(date.getDate()).padStart(2, '0');
  const hours = String(date.getHours()).padStart(2, '0');
  const minutes = String(date.getMinutes()).padStart(2, '0');
  const seconds = String(date.getSeconds()).padStart(2, '0');
  
  // Format in ISO format without timezone information
  return `${year}-${month}-${day}T${hours}:${minutes}:${seconds}`;
};

/**
 * Prepares date for API request to prevent timezone issues
 * @param {Object} data - The data object containing scheduled_date
 * @returns {Object} Processed data with corrected date format
 */
export const prepareScheduledDate = (data) => {
  if (!data || !data.scheduled_date) return data;
  
  const processed = { ...data };
  const dateValue = processed.scheduled_date;
  
  // Handle Date objects
  if (dateValue instanceof Date) {
    // Extract time from scheduled_time if provided (for estimates)
    if (processed.schedule_time) {
      try {
        // Parse the time string
        const timeValue = processed.schedule_time;
        let hours = 0;
        let minutes = 0;
        
        // Handle different time formats
        if (timeValue.includes(':')) {
          const [hourStr, minuteStr] = timeValue.split(':');
          const isPM = minuteStr.includes('PM');
          
          // Extract hours (handling 12-hour format)
          hours = parseInt(hourStr, 10);
          if (isPM && hours < 12) hours += 12;
          if (!isPM && hours === 12) hours = 0;
          
          // Extract minutes
          minutes = parseInt(minuteStr.replace(/\s?[AP]M/, ''), 10);
        } else if (timeValue.includes('AM') || timeValue.includes('PM')) {
          // Handle "1 PM" format
          const isPM = timeValue.includes('PM');
          hours = parseInt(timeValue, 10);
          if (isPM && hours < 12) hours += 12;
          if (!isPM && hours === 12) hours = 0;
        }
        
        // Create a new date with the specified time
        const updatedDate = new Date(dateValue);
        updatedDate.setHours(hours, minutes, 0, 0);
        
        // Use the local ISO string format to preserve exact time
        processed.scheduled_date = toLocalISOString(updatedDate);
      } catch (error) {
        console.error('Error parsing schedule_time:', error);
        processed.scheduled_date = toLocalISOString(dateValue);
      }
    } else {
      // For job scheduling, remove time component
      if (processed.jobScheduling) {
        const year = dateValue.getFullYear();
        const month = String(dateValue.getMonth() + 1).padStart(2, '0');
        const day = String(dateValue.getDate()).padStart(2, '0');
        processed.scheduled_date = `${year}-${month}-${day}`;
      } else {
        // For estimates, preserve the exact time using local ISO string
        processed.scheduled_date = toLocalISOString(dateValue);
      }
    }
  } 
  // Handle ISO strings with timezone (like from date pickers)
  else if (typeof dateValue === 'string' && 
          (dateValue.endsWith('Z') || dateValue.includes('+') || dateValue.includes('-', 10))) {
    
    // Parse the string to a Date object to get local components
    const parsedDate = new Date(dateValue);
    
    // For job scheduling, use date only
    if (processed.jobScheduling) {
      const year = parsedDate.getFullYear();
      const month = String(parsedDate.getMonth() + 1).padStart(2, '0');
      const day = String(parsedDate.getDate()).padStart(2, '0');
      processed.scheduled_date = `${year}-${month}-${day}`;
    } else {
      // For estimates, preserve the exact time
      processed.scheduled_date = toLocalISOString(parsedDate);
    }
  }
  
  return processed;
};

/**
 * Formats a date for display in the UI
 * @param {string|Date} date - The date to format
 * @param {Object} options - Format options
 * @returns {string} Formatted date string
 */
export const formatDateForDisplay = (date, options = {}) => {
  if (!date) return '';
  
  const { includeTime = false, includeDay = false } = options;
  
  try {
    // Convert to Date object if it's a string
    const dateObj = typeof date === 'string' ? new Date(date) : date;
    
    if (!(dateObj instanceof Date) || isNaN(dateObj.getTime())) {
      return '';
    }
    
    // Format options
    const formatOptions = {
      year: 'numeric',
      month: 'short',
      day: 'numeric',
    };
    
    if (includeDay) {
      formatOptions.weekday = 'short';
    }
    
    if (includeTime) {
      formatOptions.hour = 'numeric';
      formatOptions.minute = '2-digit';
      formatOptions.hour12 = true;
    }
    
    return new Intl.DateTimeFormat('en-US', formatOptions).format(dateObj);
  } catch (error) {
    console.error('Error formatting date for display:', error);
    return '';
  }
};

File: backend/services/date_utils.py
# backend/services/date_utils.py
import re
import pytz
import logging
from datetime import datetime, date
from models import Job

logger = logging.getLogger(__name__)

# Configure server timezone
SERVER_TIMEZONE = pytz.timezone('America/Los_Angeles')

def format_date_for_response(date_obj):
    """
    Format a date object for consistent API responses.
    For job scheduling, we want to preserve the exact date (without time) to
    prevent timezone conversion issues.
    
    Args:
        date_obj (date or datetime): The date to format
        
    Returns:
        str: Formatted date string in ISO format
    """
    if not date_obj:
        return None
        
    if isinstance(date_obj, datetime):
        # Make datetime timezone-aware if it isn't already
        if date_obj.tzinfo is None:
            date_obj = SERVER_TIMEZONE.localize(date_obj)
        # For jobs, we just want the date part in YYYY-MM-DD format
        return date_obj.date().isoformat()
    elif isinstance(date_obj, date):
        # Already just a date
        return date_obj.isoformat()
    
    # Fallback - shouldn't reach here
    return str(date_obj)

def parse_job_date(date_str):
    """
    Parse a date string for job scheduling, returning a date object
    without time component to prevent timezone issues.
    
    Args:
        date_str (str): Date string to parse
        
    Returns:
        date: Parsed date object (without time)
    """
    logger.info(f"Parsing job date: '{date_str}'")
    
    if not date_str:
        return None
    
    try:
        # Case 1: ISO format with timezone
        if 'T' in date_str and (date_str.endswith('Z') or '+' in date_str or '-' in date_str.split('T')[1]):
            logger.info(f"Parsing job date with timezone: {date_str}")
            # Parse the date and extract just the date part
            dt = datetime.fromisoformat(date_str.replace('Z', '+00:00'))
            # Convert to server timezone
            dt = dt.astimezone(SERVER_TIMEZONE)
            logger.info(f"Extracted date part: {dt.date()}")
            return dt.date()
            
        # Case 2: ISO format without timezone (2023-05-21T10:00:00)
        elif 'T' in date_str:
            logger.info(f"Parsing job date with time but no timezone: {date_str}")
            # Just parse the date part
            date_part = date_str.split('T')[0]
            year, month, day = map(int, date_part.split('-'))
            return date(year, month, day)
            
        # Case 3: Simple date format (2023-05-21)
        elif date_str.count('-') == 2:
            logger.info(f"Parsing simple YYYY-MM-DD format: {date_str}")
            year, month, day = map(int, date_str.split('-'))
            return date(year, month, day)
            
        # Case 4: Other date formats
        else:
            logger.info(f"Attempting to parse with datetime: {date_str}")
            # Try standard parsing but extract just the date part
            dt = datetime.fromisoformat(date_str)
            return dt.date()
            
    except Exception as e:
        logger.error(f"Error parsing job date '{date_str}': {str(e)}")
        raise ValueError(f"Invalid date format: {str(e)}")

def parse_datetime_with_logging(date_str, hour=None, minute=None):
    """
    Parse date string with detailed logging to diagnose timezone issues.
    
    Args:
        date_str (str): Date string to parse
        hour (int, optional): Hour value if provided separately
        minute (int, optional): Minute value if provided separately
        
    Returns:
        datetime: Parsed datetime object preserving the intended time
    """
    logger.info(f"Parsing date string: '{date_str}', hour: {hour}, minute: {minute}")
    
    if not date_str:
        logger.warning("Empty date string provided")
        return None
    
    try:
        # Case 1: ISO format with timezone information
        if ('T' in date_str and 
            (date_str.endswith('Z') or '+' in date_str or '-' in date_str.split('T')[1])):
            
            logger.info(f"Detected ISO format with timezone: {date_str}")
            
            # Replace Z with +00:00 for ISO format compatibility
            if date_str.endswith('Z'):
                date_str = date_str[:-1] + '+00:00'
                logger.info(f"Replaced Z with +00:00: {date_str}")
            
            # Parse ISO format with timezone
            dt = datetime.fromisoformat(date_str)
            logger.info(f"Parsed with timezone: {dt} (UTC)")
            
            # Convert to server timezone to preserve the time
            local_dt = dt.astimezone(SERVER_TIMEZONE)
            logger.info(f"Converted to server timezone: {local_dt} ({SERVER_TIMEZONE})")
            
            return local_dt
        
        # Case 2: ISO format without timezone (2023-05-21T10:00:00)
        elif 'T' in date_str:
            logger.info(f"Detected ISO format without timezone: {date_str}")
            
            # Parse the datetime components
            date_part, time_part = date_str.split('T')
            year, month, day = map(int, date_part.split('-'))
            
            # Parse time with proper handling of seconds/milliseconds
            time_parts = time_part.split(':')
            hour_val = int(time_parts[0])
            minute_val = int(time_parts[1])
            
            # Handle seconds if present
            second_val = 0
            if len(time_parts) > 2:
                # Remove milliseconds if present
                seconds_part = time_parts[2].split('.')[0]
                second_val = int(seconds_part)
            
            # Create datetime with the exact time specified
            naive_dt = datetime(year, month, day, hour_val, minute_val, second_val)
            logger.info(f"Created naive datetime: {naive_dt}")
            
            # Localize to server timezone
            local_dt = SERVER_TIMEZONE.localize(naive_dt)
            logger.info(f"Localized to server timezone: {local_dt}")
            
            return local_dt
        
        # Case 3: Simple date format (2023-05-21) with explicit time parameters
        elif re.match(r'^\d{4}-\d{2}-\d{2}$', date_str):
            logger.info(f"Detected date-only format: {date_str}")
            
            year, month, day = map(int, date_str.split('-'))
            
            # Use provided hour/minute or default to noon
            hour_val = hour if hour is not None else 12
            minute_val = minute if minute is not None else 0
            
            logger.info(f"Using time components: hour={hour_val}, minute={minute_val}")
            
            # Create datetime with specified or default time
            naive_dt = datetime(year, month, day, hour_val, minute_val, 0)
            logger.info(f"Created naive datetime with specified time: {naive_dt}")
            
            # Localize to server timezone
            local_dt = SERVER_TIMEZONE.localize(naive_dt)
            logger.info(f"Localized to server timezone: {local_dt}")
            
            return local_dt
            
        # Case 4: Any other format - try standard parsing
        else:
            logger.warning(f"Unrecognized date format: {date_str}, attempting standard parsing")
            dt = datetime.fromisoformat(date_str)
            logger.info(f"Parsed with standard method: {dt}")
            return dt
            
    except Exception as e:
        logger.error(f"Error parsing date '{date_str}': {str(e)}")
        raise ValueError(f"Invalid date format: {str(e)}")

def format_datetime_for_response(dt):
    """
    Format datetime for consistent API responses.
    
    Args:
        dt (datetime): Datetime object to format
        
    Returns:
        str: Formatted datetime string
    """
    if not dt:
        return None
        
    # Ensure timezone awareness
    if dt.tzinfo is None:
        dt = SERVER_TIMEZONE.localize(dt)
        
    # ISO format with timezone information
    iso_format = dt.isoformat()
    logger.info(f"Formatted datetime for response: {iso_format}")
    return iso_format

def parse_time_string(time_str):
    """
    Parse a time string in various formats and return hour and minute.
    
    Args:
        time_str (str): Time string in formats like "1:30 PM", "13:30", "1 PM"
        
    Returns:
        tuple: (hour, minute) as integers
    """
    if not time_str:
        return None, None
        
    try:
        # Normalize and clean the string
        time_str = time_str.strip().upper()
        is_pm = 'PM' in time_str
        is_am = 'AM' in time_str
        
        # Remove AM/PM indicators
        time_str = time_str.replace('AM', '').replace('PM', '').strip()
        
        # Parse hour and minute
        if ':' in time_str:
            # Format with hours and minutes (e.g., "1:30")
            hour_str, minute_str = time_str.split(':')
            hour = int(hour_str)
            minute = int(minute_str)
        else:
            # Format with just hours (e.g., "1")
            hour = int(time_str)
            minute = 0
            
        # Convert from 12-hour to 24-hour format if needed
        if is_pm and hour < 12:
            hour += 12
        if is_am and hour == 12:
            hour = 0
            
        logger.info(f"Parsed time string '{time_str}' to hour={hour}, minute={minute}")
        return hour, minute
    
    except Exception as e:
        logger.error(f"Error parsing time string '{time_str}': {str(e)}")
        return None, None

def generate_job_number():
    """Generate a unique job number based on current date and count"""
    today = date.today()
    
    # Use the specific 2-letter month codes
    month_codes = {
        1: "JA", 2: "FB", 3: "MR", 4: "AP", 5: "MY", 6: "JU",
        7: "JL", 8: "AG", 9: "SP", 10: "OT", 11: "NV", 12: "DC"
    }
    month_code = month_codes[today.month]
    
    # Get count of jobs created this month (INCLUDING today)
    month_start = date(today.year, today.month, 1)
    month_end = date(today.year, today.month + 1, 1) if today.month < 12 else date(today.year + 1, 1, 1)
    
    # Count ALL jobs in current month, including today
    month_jobs = Job.query.filter(
        Job.created_at >= month_start,
        Job.created_at < month_end
    ).count()
    
    # Add 1 to the count for the new job number
    job_number = f"{month_code}{month_jobs + 1}{str(today.year)[2:]}"
    return job_number